[Model, Core] Support Granite Speech & LoRA for STT

alex-jw-brooks opened the pull request:
<!-- markdownlint-disable -->
## Purpose
This PR adds compatibility with transcriptions / translations for granite speech! It also passes the LoRA request through to the underlying generate calls, which is needed for this model since granite speech always should have its colocated LoRA active when processing audio.

## Test Plan
I added a test to the entrypoint for transcriptions for checking this model w/ LoRA active (which also should suffice for translations since they're both using the same STT component where the lora is passed).

Can also be validated with the following:

Start server
```bash
python3 -m vllm.entrypoints.openai.api_server \
    --model ibm-granite/granite-speech-3.3-2b \
    --max-model-len 2048 \
    --enable-lora  \
    --lora-modules speech=ibm-granite/granite-speech-3.3-2b \
    --enforce-eager \
    --max-lora-rank 64
```

Example client for transcribe:
```python
# SPDX-License-Identifier: Apache-2.0
import asyncio
import json

import httpx
from openai import OpenAI

from vllm.assets.audio import AudioAsset

mary_had_lamb = AudioAsset('mary_had_lamb').get_local_path()
winning_call = AudioAsset('winning_call').get_local_path()
lora_model_name = "speech"

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "token-abc123"
openai_api_base = "http://localhost:8000/v1"
client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)


def sync_openai():
    with open(str(mary_had_lamb), "rb") as f:
        transcription = client.audio.transcriptions.create(
            file=f,
            model=lora_model_name, ### <<<< should NOT be the base model name
            language="en",
            response_format="json",
            temperature=0.0)
        print("transcription result:", transcription)


sync_openai()

# OpenAI Transcription API client does not support streaming.
async def stream_openai_response():
    data = {
        "language": "en",
        "stream": True,
        "model": lora_model_name, ### <<<< should NOT be the base model name
        "temperature": 0,
    }
    url = openai_api_base + "/audio/transcriptions"
    print("transcription result:", end=' ')
    async with httpx.AsyncClient() as client:
        with open(str(mary_had_lamb), "rb") as f:
            async with client.stream('POST', url, files={'file': f},
                                     data=data) as response:
                async for line in response.aiter_lines():
                    # Each line is a JSON object prefixed with 'data: '
                    if line:
                        if line.startswith('data: '):
                            line = line[len('data: '):]
                        # Last chunk, stream ends
                        if line.strip() == '[DONE]':
                            break
                        # Parse the JSON response
                        chunk = json.loads(line)
                        # Extract and print the content
                        content = chunk['choices'][0].get('delta',
                                                          {}).get('content')
                        print(content, end='')


# Run the asynchronous function
asyncio.run(stream_openai_response())
```

Example client for translate:
```python
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
import asyncio
import json

import httpx
from openai import OpenAI

from vllm.assets.audio import AudioAsset
lora_model_name = "speech"

def sync_openai(audio_path: str, client: OpenAI):
    with open(audio_path, "rb") as f:
        translation = client.audio.translations.create(
            file=f,
            model=lora_model_name,
            response_format="json",
            temperature=0.0,
            # Additional params not provided by OpenAI API.
            extra_body=dict(
                language="en",
                to_language="es",
                seed=4419,
                repetition_penalty=1.3,
            ),
        )
        print("translation result:", translation.text)


async def stream_openai_response(audio_path: str, base_url: str, api_key: str):
    data = {
        "language":"en",
        "to_language":"es",
        "stream": True,
        "model": lora_model_name,
    }
    url = base_url + "/audio/translations"
    headers = {"Authorization": f"Bearer {api_key}"}
    print("translation result:", end=" ")
    # OpenAI translation API client does not support streaming.
    async with httpx.AsyncClient() as client:
        with open(audio_path, "rb") as f:
            async with client.stream(
                "POST", url, files={"file": f}, data=data, headers=headers
            ) as response:
                async for line in response.aiter_lines():
                    # Each line is a JSON object prefixed with 'data: '
                    if line:
                        if line.startswith("data: "):
                            line = line[len("data: ") :]
                        # Last chunk, stream ends
                        if line.strip() == "[DONE]":
                            break
                        # Parse the JSON response
                        chunk = json.loads(line)
                        # Extract and print the content
                        content = chunk["choices"][0].get("delta", {}).get("content")
                        print(content, end="")


def main():
    mary_had_lamb = AudioAsset('mary_had_lamb').get_local_path()


    # Modify OpenAI's API key and API base to use vLLM's API server.
    openai_api_key = "EMPTY"
    openai_api_key = "token-abc123"

    openai_api_base = "http://localhost:8000/v1"
    client = OpenAI(
        api_key=openai_api_key,
        base_url=openai_api_base,
    )
    sync_openai(mary_had_lamb, client)
    # Run the asynchronous function
    asyncio.run(stream_openai_response(mary_had_lamb, openai_api_base, openai_api_key))


if __name__ == "__main__":
    main()
```

## Test Result
Test should pass, and for the respective snippets above we should get the following:

Transcription result produces:
```
transcription result: Transcription(text='the first words i spoke in the original phonograph a little piece of practical poetry mary had a little lamb it smelt like white as snow and everywhere that mary went the lamb was sure to go', logprobs=None, usage=UsageDuration(seconds=16.0, type='duration'))
transcription result: the first words i spoke in the original phonograph a little piece of practical poetry mary had a little lamb it smelt like white as snow and everywhere that mary went the lamb was sure to go
```

Translation result produces (note that this is maty had a little lamb -> Spanish since this model doesn't officially support Italian):
```
translation result: las primeras palabras que hablé en el original son un poco de poesía práctica: mary tenía un pequeño cordero, que brillaba como la nieve, y dondequiera que mary fue, el cordero seguía a ella.
translation result: las primeras palabras que hablé en el original son un poco de poesía práctica: mary tenía un pequeño cordero, que brillaba como la nieve, y dondequiera que mary fue, el cordero seguía a ella.
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

alex-jw-brooks committed:
First try at transcription support

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
pass lora requests through for transcription

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Support translations api

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Fix sad case

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add test for transcribe lora

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

DarkLight1337 reviewed (commented):
cc @jeejeelee @NickLucche

NickLucche commented on tests/entrypoints/openai/test_transcription_validation.py:
```diff
@@ -60,6 +60,38 @@ async def test_basic_audio(mary_had_lamb, model_name):
 
 
+@pytest.mark.asyncio
+async def test_basic_audio_with_lora(mary_had_lamb):
+    """Ensure STT requests can pass LoRA requests through to generate."""
+    model_name = "ibm-granite/granite-speech-3.3-2b"
```
do you want to replicate this test by checking translation in `test_translation_validation`?

NickLucche reviewed (approved):
Hey Alex thank you for this PR!
I think current PR is fine as is (only left one comment), but it would be nice (in this or in a following contrib) to test granite on other unit tests (eg `test_long_audio_request`); possibly expanding the `server()` fixture.

NickLucche commented:
Almost forgot, can we update docs https://docs.vllm.ai/en/stable/models/supported_models.html#transcription with Granite Speech?

alex-jw-brooks committed:
Add translate test

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Move json / assert stuff out of ctx mgr

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
Add granite speech to transcription models

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
try reducing max num seqs

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
run code formatting

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks committed:
change test order to avoid oomkill

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks commented:
Hey @NickLucche, just wanted to follow up on this PR, sorry for the delays! Could you please take another look when you can? I've added the translation test and fixed the CI 🙂

NickLucche reviewed (commented):
Thanks for the work @alex-jw-brooks , only left a minor comment, o/w it's lgtm. 
Can't say I am familiar with the llm.py lora-related change though

alex-jw-brooks committed:
move lora req into preprocess err handling

Signed-off-by: Alex-Brooks <Alex.Brooks@ibm.com>

alex-jw-brooks commented on vllm/entrypoints/llm.py:
```diff
@@ -428,7 +428,32 @@ def generate(
         # Add any modality specific loras to the corresponding prompts
-        lora_request = self._get_modality_specific_lora_reqs(prompts, lora_request)
+        user_lora_request = lora_request
+        lora_request = self._get_modality_specific_lora_reqs(prompts, user_lora_request)
```
@NickLucche this actually shouldn't be in this PR, sorry about that! 🤦‍♂️that change was for this [issue](https://github.com/vllm-project/vllm/issues/26157), I accidentally included it while rebasing.

I've removed the llm changes here, but will copy you on the other PR when it's open for awareness 🙂

alex-jw-brooks commented on tests/entrypoints/openai/test_translation_validation.py:
```diff
@@ -48,6 +48,40 @@ async def test_non_asr_model(foscolo):
+            temperature=0.0,
+        )
+    out = json.loads(translation)["text"].strip().lower()
+    assert "mary tenía un pequeño cordero" in out
```
I opened a PR to loosen the check to just look for pequeño, which would at least let the above build pass here https://github.com/vllm-project/vllm/pull/28247.

Hopefully this will fix the flakiness in current builds, if it continues to be flaky, happy to help investigate as well
